File: blanc/src/algorithm/leetcode/p201to250/LeetCode210CourseScheduleII.java
package algorithm.leetcode.p201to250;

import java.util.*;

@SuppressWarnings("unused")
public class LeetCode210CourseScheduleII {

    private interface Method {
        int[] findOrder(int numCourses, int[][] prerequisites);
    }

    private static class BFS implements Method {

        @SuppressWarnings("unchecked")
        @Override
        public int[] findOrder(int numCourses, int[][] prerequisites) {
            int index = 0;
            int[] res = new int[numCourses];
            int[] inDegrees = new int[numCourses];
            List<Integer>[] successorsArray = new List[numCourses];
            Queue<Integer> taskQueue = new ArrayDeque<>();

            for (int i = 0; i < numCourses; i++) {
                successorsArray[i] = new ArrayList<>(0);
            }
            for (int[] prerequisite : prerequisites) {
                inDegrees[prerequisite[0]]++;
                successorsArray[prerequisite[1]].add(prerequisite[0]);
            }
            for (int i = 0; i < numCourses; i++) {
                if (inDegrees[i] == 0) {
                    taskQueue.offer(i);
                }
            }

            while (!taskQueue.isEmpty()) {
                int course = taskQueue.poll();
                res[index++] = course;
                for (int successor : successorsArray[course]) {
                    if (--inDegrees[successor] == 0) {
                        taskQueue.offer(successor);
                    }
                }
            }

            return index == numCourses ? res : new int[0];
        }
    }
}

File: src/leetcode/p201to250/LeetCode210CourseScheduleII.kt
package leetcode.p201to250

import java.util.*

class LeetCode210CourseScheduleII {

    fun findOrder(numCourses: Int, prerequisites: Array<IntArray>?): IntArray {
        if (numCourses == 0 || prerequisites == null) {
            return IntArray(0)
        }
        var index = 0
        val res = IntArray(numCourses)
        val inDegrees = IntArray(numCourses)
        val successors = HashMap<Int, MutableList<Int>>()
        val taskList = ArrayDeque<Int>()
        for (prerequisite in prerequisites) {
            inDegrees[prerequisite[0]]++
            successors.getOrPut(prerequisite[1]) { mutableListOf() }.add(prerequisite[0])
        }
        for (i in 0 until numCourses) {
            if (inDegrees[i] == 0) {
                taskList.offer(i)
            }
        }
        while (!taskList.isEmpty()) {
            val cursor = taskList.poll()
            res[index++] = cursor
            for (successor in successors[cursor] ?: emptyList<Int>()) {
                if (--inDegrees[successor] == 0) {
                    taskList.offer(successor)
                }
            }
        }
        return if (index == numCourses) res else IntArray(0)
    }
}
